### ISSUE TITLE
Project 2: Level 6 Human server and self-checkout common parent a must?

### ISSUE THREAD

**Issue by jadenwjh** [nus-cs2030/2021-s1#434]
In the question description it was stated:

Two types of servers,
human servers who may rest after serving a customer, and
self-checkout counters that never rest

On surface it will make sense that the question is hinting to make both classes extend from a common parent. But the two classes have v different implementations. Do we need to force self checkout and human server to have a common parent?

**Comment by jiayi-lim** [comment]
I think it will be good to have a common parent since both have similar attributes such as - availability, next available time, the number of customers in the queue, queue max length. If you make the parent class, it will reflect good coding practices but it should work alright without a parent class!

**Comment by SkyBlaise99** [comment]
It gets messy if you are to revamp all over, especially at this later stage, you might end up debugging your newly created parent class. So perhaps adding new variable to existing `Server.java` is a better approach. Tbh, you just need 2 extra variables iirc

**Comment by chongweil907** [comment]
@SkyBlaise99 Can u elaborate abit more on this ?

**Comment by descentbw** [comment]
I didn't make a common parent and my code works fine. Simply add more attributes to server, like the customers in the queue. That being said, having a common parent class is fine too.

**Comment by jeremybrett4869** [comment]
For me, I create a parent class for the servers and self check-out counters to prevent duplicate codes as they can inherit the attributes and methods from the parent class. But I think it should also work without the parent class.

**Comment by kheekheekhee** [comment]
I didn't make a separate class for self-checkout servers since they function the exact same way. I simply added a attribute called serverType in my constructor that I can use to check what kind of Server it is

**Comment by SkyBlaise99** [comment]
You can have something to denote the type of server and a static variable to store the shared queue for the self service server. Then depending on the type of server, you can either use shared queue or normal queue

**Comment by kaiyang96** [comment]
I made my self-checkout servers simply a subclass of Server without any additional properties or methods. My Shop class was the one that implemented and managed all the queues and other behaviours involving the 2 different kinds of servers.

**Comment by roopaniedu** [comment]
@kaiyang96 I'm planning to do the same as you. However, i am confused regarding the extra functionalities Shop.java needs. How does the queue work for a selfcheckout class that is a subclass of Server? (my Server has a list queue that holds all the customers waiting in level 5)

**Comment by weiyifei-evelyn** [comment]
No, it is not a must. I was stuck at this question also. Some suggest that I can just use a property called `isSelfService` to label the server as a self-checkout counter. Also, it is not necessary to have a static queue shared. You can just use a static number to denote the number of waiting customers in the queue shared.

**Comment by rcdl2222** [comment]
@bitz-nobel I have the same design in which the SelfCheckoutCounter is a subclass of Server, giving it additional properties (such as not needing to rest etc). As of the queue, I made used of a static variable similar to @SkyBlaise99 , so that it is easier for the SelfCheckoutCounters to retrieve the customers. 

The above design will work if your Shop stores Servers, as SelfCheckoutCounter will be a subclass of Server.

**Comment by TarciusW** [comment]
it is not compulsory to have a common parent for human and self servers but do keep in mind that you will have to tweak your shop.java as required depending. both ways has its perks, so its really dependent on how much you want to tweak your code at this point in the project.
